### PR TITLE
Tiffany/gas pricing

### PR DESCRIPTION
**Pull request by tmckenzie51**
### What was wrong?

> construct_time_based_gas_price_strategy is unable to handle cases when network suddenly requires higher gas prices due to many variables such as network congestion. This has resulted in more transactions to revert.

Related to Issue #1463

### How was it fixed?
A recency based weighted average of the block time is calculated. This weighted_avg_block_time is then used to calculate the probabilities required to determine the best gas price. 

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by tmckenzie51** [comment]
Hi there! @jmjin @kclowes 

I just created this weighted average function that uses the weighted average by recency of the block in the sample, in order to calculate the weighted average block time for the time_based_gas_price_strategy function.  I used n/N (where N is the sample size and n is the number of the block in the sample 1 through to N, with the Nth block being equivalent to the 'latest' block) as a simple weighting function that could capture the idea of weighting more recent blocks more heavily. Do you think that such a weighting function would achieve the goal of more accurately deciding a gas price during a time of sudden network congestion?

File: .idea/codeStyles/codeStyleConfig.xml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<component name="ProjectCodeStyleConfiguration">
```
Comment (kclowes) [inline .idea/codeStyles/codeStyleConfig.xml]: This doesn't need to be added to the codebase :) may want to add it to our gitignore though

**Comment by tmckenzie51** [comment]
> This looks like the right approach to me @tmckenzie51! Thank you! I think this deserves to be it's own strategy, meaning that I think we should leave the simple time based strategy, and have a separate weighted strategy. I'd like to see some more testing around this strategy to make sure it's working as expected, and there is also some white space that needs to be cleaned up. I'll plan on addressing this feedback since I'm sure you are busy with other things now :)

@kclowes That makes sense to have it as a separate strategy. Thanks for pointing that out, and thank you for the feedback and for handling this for me. :)

File: web3/gas_strategies/time_based.py [hunk 2]
```diff
@@ -55,6 +54,27 @@ def _get_avg_block_time(w3: Web3, sample_size: int) -> float:
     return (latest['timestamp'] - oldest['timestamp']) / constrained_sample_size
 
 
+def _get_weighted_avg_block_time(w3: Web3, sample_size: int) -> float:
```
Comment (kclowes) [inline web3/gas_strategies/time_based.py:57]: @pipermerriam do you have ideas on how we can test this? I did manually test it and confirm it works, but the eth-tester block times are all the same AFAICT.

**Comment by kclowes** [comment]
@marcgarreau I think this is good to go if you want to take a look and give final approval!

**Pull request merged**